File: tests/wasm/test_frama.js
/**
 * WASM binding tests for FRAMA indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('FRAMA partial params', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.frama_js(high, low, close, 10, 300, 1);
    assert.strictEqual(result.length, close.length);
    
    
    const resultCustomWindow = wasm.frama_js(high, low, close, 14, 300, 1);
    assert.strictEqual(resultCustomWindow.length, close.length);
    
    const resultCustomSc = wasm.frama_js(high, low, close, 10, 200, 1);
    assert.strictEqual(resultCustomSc.length, close.length);
    
    const resultCustomFc = wasm.frama_js(high, low, close, 10, 300, 2);
    assert.strictEqual(resultCustomFc.length, close.length);
});

test('FRAMA accuracy', async () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    const expectedLast5 = [
        59337.23056930512,
        59321.607512374605,
        59286.677929994796,
        59268.00202402624,
        59160.03888720062
    ];
    
    const result = wasm.frama_js(high, low, close, 10, 300, 1);
    
    assert.strictEqual(result.length, close.length);
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expectedLast5,
        0.1,  
        "FRAMA last 5 values mismatch"
    );
    
    
    await compareWithRust('frama', result, 'high,low,close', {window: 10, sc: 300, fc: 1});
});

test('FRAMA empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.frama_js(empty, empty, empty, 10, 300, 1);
    }, /Input data slice is empty/);
});

test('FRAMA zero window', () => {
    
    const high = new Float64Array([10.0, 20.0, 30.0]);
    const low = new Float64Array([5.0, 15.0, 25.0]);
    const close = new Float64Array([7.0, 17.0, 27.0]);
    
    assert.throws(() => {
        wasm.frama_js(high, low, close, 0, 300, 1);
    }, /Invalid window/);
});

test('FRAMA window exceeds length', () => {
    
    const high = new Float64Array([10.0, 20.0, 30.0]);
    const low = new Float64Array([5.0, 15.0, 25.0]);
    const close = new Float64Array([7.0, 17.0, 27.0]);
    
    assert.throws(() => {
        wasm.frama_js(high, low, close, 10, 300, 1);
    }, /Invalid window/);
});

test('FRAMA very small dataset', () => {
    
    const high = new Float64Array([42.0]);
    const low = new Float64Array([40.0]);
    const close = new Float64Array([41.0]);
    
    assert.throws(() => {
        wasm.frama_js(high, low, close, 10, 300, 1);
    }, /Invalid window|Not enough valid data/);
});

test('FRAMA mismatched lengths', () => {
    
    const high = new Float64Array([1.0, 2.0, 3.0]);
    const low = new Float64Array([0.5, 1.5]);
    const close = new Float64Array([1.0]);
    
    assert.throws(() => {
        wasm.frama_js(high, low, close, 10, 300, 1);
    }, /Mismatched slice lengths/);
});

test('FRAMA all NaN input', () => {
    
    const allNaN = new Float64Array(10);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.frama_js(allNaN, allNaN, allNaN, 10, 300, 1);
    }, /All values are NaN/);
});

test('FRAMA not enough valid data', () => {
    
    const high = new Float64Array([NaN, NaN, 10.0, 20.0, 30.0]);
    const low = new Float64Array([NaN, NaN, 5.0, 15.0, 25.0]);
    const close = new Float64Array([NaN, NaN, 7.0, 17.0, 27.0]);
    
    
    assert.throws(() => {
        wasm.frama_js(high, low, close, 10, 300, 1);
    }, /Invalid window/);
    
    
    const high2 = new Float64Array([NaN, NaN, NaN, NaN, NaN, 10.0, 20.0, 30.0, 40.0, 50.0]);
    const low2 = new Float64Array([NaN, NaN, NaN, NaN, NaN, 5.0, 15.0, 25.0, 35.0, 45.0]);
    const close2 = new Float64Array([NaN, NaN, NaN, NaN, NaN, 7.0, 17.0, 27.0, 37.0, 47.0]);
    
    
    assert.throws(() => {
        wasm.frama_js(high2, low2, close2, 10, 300, 1);
    }, /Not enough valid data/);
});

test('FRAMA NaN handling', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    
    const result = wasm.frama_js(high, low, close, 10, 300, 1);
    assert.strictEqual(result.length, close.length);
    
    
    assertAllNaN(result.slice(0, 9), "Expected NaN in warmup period");
    
    
    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
});

test('FRAMA batch single parameter set', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    
    
    const batchResult = wasm.frama_batch_js(
        high, low, close,
        10, 10, 0,    
        300, 300, 0,  
        1, 1, 0       
    );
    
    
    const singleResult = wasm.frama_js(high, low, close, 10, 300, 1);
    
    assert.strictEqual(batchResult.length, singleResult.length);
    assertArrayClose(batchResult, singleResult, 1e-10, "Batch vs single mismatch");
});

test('FRAMA batch multiple parameters', () => {
    
    const high = new Float64Array(testData.high.slice(0, 100));
    const low = new Float64Array(testData.low.slice(0, 100));
    const close = new Float64Array(testData.close.slice(0, 100));
    
    
    const batchResult = wasm.frama_batch_js(
        high, low, close,
        8, 12, 2,      
        200, 300, 100, 
        1, 2, 1        
    );
    
    
    assert.strictEqual(batchResult.length, 12 * 100);
    
    
    const firstRow = batchResult.slice(0, 100);
    const singleResult = wasm.frama_js(high, low, close, 8, 200, 1);
    assertArrayClose(
        firstRow, 
        singleResult, 
        1e-10, 
        "First batch row mismatch"
    );
});

test('FRAMA batch metadata', () => {
    
    const metadata = wasm.frama_batch_metadata_js(
        8, 12, 2,      
        200, 300, 100, 
        1, 2, 1        
    );
    
    
    assert.strictEqual(metadata.length, 12 * 3);
    
    
    assert.strictEqual(metadata[0], 8);    
    assert.strictEqual(metadata[1], 200);  
    assert.strictEqual(metadata[2], 1);    
    
    
    assert.strictEqual(metadata[3], 8);    
    assert.strictEqual(metadata[4], 200);  
    assert.strictEqual(metadata[5], 2);    
    
    
    assert.strictEqual(metadata[6], 8);    
    assert.strictEqual(metadata[7], 300);  
    assert.strictEqual(metadata[8], 1);    
});

test('FRAMA batch warmup validation', () => {
    
    const high = new Float64Array(testData.high.slice(0, 50));
    const low = new Float64Array(testData.low.slice(0, 50));
    const close = new Float64Array(testData.close.slice(0, 50));
    
    const batchResult = wasm.frama_batch_js(
        high, low, close,
        6, 10, 4,     
        300, 300, 0,  
        1, 1, 0       
    );
    
    const metadata = wasm.frama_batch_metadata_js(6, 10, 4, 300, 300, 0, 1, 1, 0);
    const numCombos = metadata.length / 3;
    assert.strictEqual(numCombos, 2);  
    
    
    for (let combo = 0; combo < numCombos; combo++) {
        const window = metadata[combo * 3];
        const warmup = window - 1;
        const rowStart = combo * 50;
        const rowData = batchResult.slice(rowStart, rowStart + 50);
        
        
        for (let i = 0; i < warmup; i++) {
            assert(isNaN(rowData[i]), 
                `Expected NaN at index ${i} for window=${window}`);
        }
        
        
        for (let i = warmup; i < 50; i++) {
            assert(!isNaN(rowData[i]), 
                `Unexpected NaN at index ${i} for window=${window}`);
        }
    }
});

test('FRAMA batch edge cases', () => {
    
    const size = 20;
    const high = new Float64Array(size).fill(0).map((_, i) => i + 2);
    const low = new Float64Array(size).fill(0).map((_, i) => i);
    const close = new Float64Array(size).fill(0).map((_, i) => i + 1);
    
    
    const singleBatch = wasm.frama_batch_js(
        high, low, close,
        10, 10, 1,
        300, 300, 1,
        1, 1, 1
    );
    
    assert.strictEqual(singleBatch.length, size);
    
    
    const zeroStepBatch = wasm.frama_batch_js(
        high, low, close,
        10, 10, 0,
        300, 300, 0,
        1, 1, 0
    );
    
    assert.strictEqual(zeroStepBatch.length, size); 
    
    
    assert.throws(() => {
        wasm.frama_batch_js(
            new Float64Array([]),
            new Float64Array([]),
            new Float64Array([]),
            10, 10, 0,
            300, 300, 0,
            1, 1, 0
        );
    }, /Input data slice is empty|All values are NaN/);
});

test('FRAMA batch performance test', () => {
    
    const high = new Float64Array(testData.high.slice(0, 200));
    const low = new Float64Array(testData.low.slice(0, 200));
    const close = new Float64Array(testData.close.slice(0, 200));
    
    
    const startBatch = Date.now();
    const batchResult = wasm.frama_batch_js(
        high, low, close,
        8, 12, 2,     
        200, 400, 100,
        1, 2, 1       
    );
    const batchTime = Date.now() - startBatch;
    
    
    const startSingle = Date.now();
    const singleResults = [];
    for (let window = 8; window <= 12; window += 2) {
        for (let sc = 200; sc <= 400; sc += 100) {
            for (let fc = 1; fc <= 2; fc += 1) {
                singleResults.push(...wasm.frama_js(high, low, close, window, sc, fc));
            }
        }
    }
    const singleTime = Date.now() - startSingle;
    
    
    assert.strictEqual(batchResult.length, singleResults.length);
    
    
    console.log(`  FRAMA Batch time: ${batchTime}ms, Single calls time: ${singleTime}ms`);
});



test.after(() => {
    console.log('FRAMA WASM tests completed');
});

File: tests/python/test_frama.py
"""
Python binding tests for FRAMA indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestFrama:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_frama_partial_params(self, test_data):
        """Test FRAMA with partial parameters - mirrors check_frama_partial_params"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']


        result = ta_indicators.frama(high, low, close, 10, 300, 1)
        assert len(result) == len(close)


        result = ta_indicators.frama(high, low, close, 14, 300, 1)
        assert len(result) == len(close)

        result = ta_indicators.frama(high, low, close, 10, 200, 1)
        assert len(result) == len(close)

        result = ta_indicators.frama(high, low, close, 10, 300, 2)
        assert len(result) == len(close)

    def test_frama_accuracy(self, test_data):
        """Test FRAMA matches expected values from Rust tests - mirrors check_frama_accuracy"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['frama']


        result = ta_indicators.frama(
            high, low, close,
            window=expected['default_params']['window'],
            sc=expected['default_params']['sc'],
            fc=expected['default_params']['fc']
        )

        assert len(result) == len(close)


        assert_close(
            result[-5:],
            expected['last_5_values'],
            rtol=0,
            atol=0.1,
            msg="FRAMA last 5 values mismatch"
        )


        compare_with_rust('frama', result, 'high,low,close', expected['default_params'])

    def test_frama_default_candles(self, test_data):
        """Test FRAMA with default parameters - mirrors check_frama_default_candles"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']


        result = ta_indicators.frama(high, low, close, 10, 300, 1)
        assert len(result) == len(close)

    def test_frama_zero_window(self):
        """Test FRAMA fails with zero window - mirrors check_frama_zero_window"""
        high = np.array([10.0, 20.0, 30.0])
        low = np.array([5.0, 15.0, 25.0])
        close = np.array([7.0, 17.0, 27.0])

        with pytest.raises(ValueError, match="Invalid window"):
            ta_indicators.frama(high, low, close, window=0, sc=300, fc=1)

    def test_frama_window_exceeds_length(self):
        """Test FRAMA fails when window exceeds data length - mirrors check_frama_window_exceeds_length"""
        high = np.array([10.0, 20.0, 30.0])
        low = np.array([5.0, 15.0, 25.0])
        close = np.array([7.0, 17.0, 27.0])

        with pytest.raises(ValueError, match="Invalid window"):
            ta_indicators.frama(high, low, close, window=10, sc=300, fc=1)

    def test_frama_very_small_dataset(self):
        """Test FRAMA fails with insufficient data - mirrors check_frama_very_small_dataset"""
        high = np.array([42.0])
        low = np.array([40.0])
        close = np.array([41.0])

        with pytest.raises(ValueError, match="Invalid window|Not enough valid data"):
            ta_indicators.frama(high, low, close, window=10, sc=300, fc=1)

    def test_frama_empty_input(self):
        """Test FRAMA fails with empty input - mirrors check_frama_empty_input"""
        empty = np.array([])

        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.frama(empty, empty, empty, window=10, sc=300, fc=1)

    def test_frama_mismatched_lengths(self):
        """Test FRAMA fails with mismatched input lengths - mirrors check_frama_mismatched_len"""
        high = np.array([1.0, 2.0, 3.0])
        low = np.array([0.5, 1.5])
        close = np.array([1.0])

        with pytest.raises(ValueError, match="Mismatched slice lengths"):
            ta_indicators.frama(high, low, close, window=10, sc=300, fc=1)

    def test_frama_all_nan(self):
        """Test FRAMA fails with all NaN values - mirrors check_frama_all_nan"""
        all_nan = np.full(10, np.nan)

        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.frama(all_nan, all_nan, all_nan, window=10, sc=300, fc=1)

    def test_frama_nan_handling(self, test_data):
        """Test FRAMA handles NaN values correctly - mirrors check_frama_nan_handling"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']

        result = ta_indicators.frama(high, low, close, window=10, sc=300, fc=1)
        assert len(result) == len(close)


        if len(result) > 240:
            assert not np.any(np.isnan(result[240:])), "Found unexpected NaN after warmup period"


        assert np.all(np.isnan(result[:9])), "Expected NaN in warmup period"

    def test_frama_streaming(self, test_data):
        """Test FRAMA streaming matches batch calculation - mirrors check_frama_streaming"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        window = 10
        sc = 300
        fc = 1


        batch_result = ta_indicators.frama(high, low, close, window=window, sc=sc, fc=fc)


        stream = ta_indicators.FramaStream(window=window, sc=sc, fc=fc)
        stream_values = []

        for h, l, c in zip(high, low, close):
            result = stream.update(h, l, c)
            stream_values.append(result if result is not None else np.nan)

        stream_values = np.array(stream_values)


        assert len(batch_result) == len(stream_values)


        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-9, atol=1e-9,
                        msg=f"FRAMA streaming mismatch at index {i}")

    def test_frama_batch_single_params(self, test_data):
        """Test FRAMA batch processing with single parameter set - mirrors check_batch_default_row"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']

        result = ta_indicators.frama_batch(
            high, low, close,
            window_range=(10, 10, 0),
            sc_range=(300, 300, 0),
            fc_range=(1, 1, 0)
        )

        assert 'values' in result
        assert 'windows' in result
        assert 'scs' in result
        assert 'fcs' in result


        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(close)


        default_row = result['values'][0]
        expected = EXPECTED_OUTPUTS['frama']['last_5_values']


        assert_close(
            default_row[-5:],
            expected,
            rtol=0,
            atol=0.1,
            msg="FRAMA batch default row mismatch"
        )

    def test_frama_batch_multiple_params(self, test_data):
        """Test FRAMA batch processing with multiple parameter combinations"""
        high = test_data['high'][:100]
        low = test_data['low'][:100]
        close = test_data['close'][:100]

        result = ta_indicators.frama_batch(
            high, low, close,
            window_range=(8, 12, 2),
            sc_range=(200, 300, 100),
            fc_range=(1, 2, 1)
        )


        assert result['values'].shape[0] == 12
        assert result['values'].shape[1] == 100


        assert len(result['windows']) == 12
        assert len(result['scs']) == 12
        assert len(result['fcs']) == 12


        single_result = ta_indicators.frama(high, low, close, window=8, sc=200, fc=1)
        assert_close(
            result['values'][0],
            single_result,
            rtol=1e-10,
            msg="First batch row mismatch"
        )

    def test_frama_batch_warmup_validation(self, test_data):
        """Test batch warmup period handling"""
        high = test_data['high'][:50]
        low = test_data['low'][:50]
        close = test_data['close'][:50]

        result = ta_indicators.frama_batch(
            high, low, close,
            window_range=(6, 10, 4),
            sc_range=(300, 300, 0),
            fc_range=(1, 1, 0)
        )


        assert result['values'].shape == (2, 50)



        assert np.all(np.isnan(result['values'][0][:5]))
        assert not np.any(np.isnan(result['values'][0][5:]))


        assert np.all(np.isnan(result['values'][1][:9]))
        assert not np.any(np.isnan(result['values'][1][9:]))

    def test_frama_not_enough_valid_data(self):
        """Test FRAMA fails when there's not enough valid data after NaN prefix"""
        high = np.array([np.nan, np.nan, 10.0, 20.0, 30.0])
        low = np.array([np.nan, np.nan, 5.0, 15.0, 25.0])
        close = np.array([np.nan, np.nan, 7.0, 17.0, 27.0])


        with pytest.raises(ValueError, match="Invalid window"):
            ta_indicators.frama(high, low, close, window=10, sc=300, fc=1)


        high2 = np.array([np.nan, np.nan, np.nan, np.nan, np.nan, 10.0, 20.0, 30.0, 40.0, 50.0])
        low2 = np.array([np.nan, np.nan, np.nan, np.nan, np.nan, 5.0, 15.0, 25.0, 35.0, 45.0])
        close2 = np.array([np.nan, np.nan, np.nan, np.nan, np.nan, 7.0, 17.0, 27.0, 37.0, 47.0])


        with pytest.raises(ValueError, match="Not enough valid data"):
            ta_indicators.frama(high2, low2, close2, window=10, sc=300, fc=1)


if __name__ == '__main__':
    pytest.main([__file__, '-v'])